File: src/main/java/net/ripe/logging/LoggingOutputStream.java
package net.ripe.logging;

import org.apache.log4j.Category;
import org.apache.log4j.Priority;

import java.io.IOException;
import java.io.OutputStream;

public class LoggingOutputStream extends OutputStream {
    protected static final String LINE_SEPARATOR = System.getProperty("line.separator");

    protected boolean hasBeenClosed = false;

    protected byte[] buf;

    protected int count;

    private int bufLength;

    public static final int DEFAULT_BUFFER_LENGTH = 2048;

    protected Category category;

    protected Priority priority;

    public LoggingOutputStream(Category cat, Priority priority)
        throws IllegalArgumentException {
        if (cat == null) {
            throw new IllegalArgumentException("cat == null");
        }
        if (priority == null) {
            throw new IllegalArgumentException("priority == null");
        }

        this.priority = priority;
        category = cat;
        bufLength = DEFAULT_BUFFER_LENGTH;
        buf = new byte[DEFAULT_BUFFER_LENGTH];
        count = 0;
    }

    public void close() {
        flush();
        hasBeenClosed = true;
    }

    public void write(final int b) throws IOException {
        if (hasBeenClosed) {
            throw new IOException("The stream has been closed.");
        }

        // don't log nulls
        if (b == 0) {
            return;
        }

        // would this be writing past the buffer?
        if (count == bufLength) {
            // grow the buffer
            final int newBufLength = bufLength + DEFAULT_BUFFER_LENGTH;
            final byte[] newBuf = new byte[newBufLength];

            System.arraycopy(buf, 0, newBuf, 0, bufLength);

            buf = newBuf;
            bufLength = newBufLength;
        }

        buf[count] = (byte) b;
        count++;
    }

    public void flush() {
        if (count == 0) {
            return;
        }

        // don't print out blank lines; flushing from PrintStream puts out these
        if (count == LINE_SEPARATOR.length()) {
            if (((char) buf[0]) == LINE_SEPARATOR.charAt(0) &&
                ((count == 1) ||  // <- Unix & Mac, -> Windows
                    ((count == 2) && ((char) buf[1]) == LINE_SEPARATOR.charAt(1)))) {
                reset();
                return;
            }
        }

        final byte[] theBytes = new byte[count];

        System.arraycopy(buf, 0, theBytes, 0, count);

        category.log(priority, new String(theBytes));

        reset();
    }

    private void reset() {
        // not resetting the buffer -- assuming that if it grew that it
        //   will likely grow similarly again
        count = 0;
    }
}

File: src/main/scala/net/ripe/logging/LoggingOutputStream.scala
package net.ripe.logging

import java.io.{IOException, OutputStream}

import org.apache.log4j.{Priority, Category}

class LoggingOutputStream(category: Category, priority: Priority) extends OutputStream {
  private val LINE_SEPARATOR = System.getProperty("line.separator")
  private var closed = false
  private var buffer = new Array[Byte](2048)
  private var count = 0

  override def close() {
    flush()
    closed = true
  }

  @throws(classOf[IOException])
  override def write(b: Int) {
    if (closed) {
      throw new IOException("The stream has been closed!")
    }
    if (b == 0) {
      return
    }

    if (count == buffer.length) {
      // The buffer is full; grow it
      val newBuffer = new Array[Byte](2 * buffer.length)
      System.arraycopy(buffer, 0, newBuffer, 0, buffer.length)
      buffer = newBuffer
    }

    buffer(count) = b.toByte
    count += 1
  }

  override def flush() {
    if (count == 0) {
      return
    }
    // Don't print out blank lines; flushing from PrintStream puts these out
    if (!isBlankLine) category.log(priority, new String(buffer.slice(0, count)))
    reset()
  }

  private def isBlankLine = (count == LINE_SEPARATOR.length) &&
    (count == 1 && (buffer(0).toChar == LINE_SEPARATOR.charAt(0)) ||
      count == 2 && (buffer(1).toChar == LINE_SEPARATOR.charAt(1)))

  private def reset() {
    count = 0
  }
}
